[rpc][jit] support rpc_sync in TorchScript

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43046 [rpc][jit] support remote call in TorchScript
* **#43043 [rpc][jit] support rpc_sync in TorchScript**
* #43042 [rpc] rename some functions

This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

Differential Revision: [D23252039](https://our.internmc.facebook.com/intern/diff/D23252039)

wanchaol committed:
[rpc][jit] support rpc_sync in TorchScript

This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc][jit] support rpc_sync in TorchScript"


This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc][jit] support rpc_sync in TorchScript"


This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

[ghstack-poisoned]

mrshenli commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -886,6 +886,11 @@ std::shared_ptr<SugaredValue> toSugaredValue(
       py::module::import("torch.distributed.rpc").attr("rpc_async").ptr()) {
     return SpecialFormValue::create(prim::rpc_async);
+  } else if (
+      // RPC module is only avaialble  when build flag "USE_DISTRIBUTED" is on.
```
redundant space after "available"

mrshenli commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -886,6 +886,11 @@ std::shared_ptr<SugaredValue> toSugaredValue(
+  } else if (
+      // RPC module is only avaialble  when build flag "USE_DISTRIBUTED" is on.
+      obj.ptr() ==
+      py::module::import("torch.distributed.rpc").attr("rpc_sync").ptr()) {
```
is the indent here intentional?

wanchaol replied:
it's formatted by clang-format (and stay aligned with the above rpc_async op)

mrshenli commented on torch/csrc/jit/runtime/register_distributed_ops.cpp:
```diff
@@ -26,6 +26,136 @@ static auto workerInfo =
+        functionSchema,
+        userCallableStack,
+        rpcTimeout);
+    futureIValuePtr->wait();
+    auto res = futureIValuePtr->value();
```
These could throw error. Is that OK?

wanchaol replied:
good catch! handled the error.

mrshenli reviewed (approved):
LGTM! Thanks for adding this!

It will be great if we can also cover the case where user function throws, but it can be added in follow up PRs.

wanchaol committed:
Update on "[rpc][jit] support rpc_sync in TorchScript"


This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc][jit] support rpc_sync in TorchScript"


This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

Differential Revision: [D23252039](https://our.internmc.facebook.com/intern/diff/D23252039)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc][jit] support rpc_sync in TorchScript"


This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

Differential Revision: [D23252039](https://our.internmc.facebook.com/intern/diff/D23252039)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc][jit] support rpc_sync in TorchScript"


This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

Differential Revision: [D23252039](https://our.internmc.facebook.com/intern/diff/D23252039)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc][jit] support rpc_sync in TorchScript"


This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

Differential Revision: [D23252039](https://our.internmc.facebook.com/intern/diff/D23252039)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc][jit] support rpc_sync in TorchScript"


This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

Differential Revision: [D23252039](https://our.internmc.facebook.com/intern/diff/D23252039)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc][jit] support rpc_sync in TorchScript"


This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

Differential Revision: [D23252039](https://our.internmc.facebook.com/intern/diff/D23252039)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc][jit] support rpc_sync in TorchScript"


This add the support for rpc_sync in TorchScript in a way similar to
rpc_async

Differential Revision: [D23252039](https://our.internmc.facebook.com/intern/diff/D23252039)

[ghstack-poisoned]
